### ISSUE TITLE
ForeignKeyConstraint constructor doesn't initialize the object completely

### ISSUE THREAD

**Issue by morozov** [doctrine/dbal#4506]
### Bug Report

|    Q        |   A
|------------ | ------
| Version     | 2.12.1
```php
<?php

use Doctrine\DBAL\Schema\ForeignKeyConstraint;

$fk = new ForeignKeyConstraint(['user_id'], 'users', ['id']);
$fk->getLocalTableName();
```
Produces:
```
Fatal error: Uncaught Error: Call to a member function getName() on null in lib/Doctrine/DBAL/Schema/ForeignKeyConstraint.php on line 113

Error: Call to a member function getName() on null in lib/Doctrine/DBAL/Schema/ForeignKeyConstraint.php on line 113

Call Stack:
    0.0005     464176   1. {main}() test.php:0
    0.0133    3261400   2. Doctrine\DBAL\Schema\ForeignKeyConstraint->getLocalTableName() test.php:6
```

**Comment by morozov** [comment]
But once the object is completely initialized, there are potential memory leaks:
1. https://github.com/doctrine/dbal/blob/f930db20d2aa4451d3f0bef4e0a5dd33dba38c74/lib/Doctrine/DBAL/Schema/Table.php#L545-L559
2. https://github.com/doctrine/dbal/blob/f930db20d2aa4451d3f0bef4e0a5dd33dba38c74/lib/Doctrine/DBAL/Schema/Table.php#L875

**Comment by greg0ire** [comment]
Sorry, can you please explain how the memleak(s) work? I don't get it.

**Comment by morozov** [comment]
There is a circular reference between the `Table` and its `ForeignKeyConstraint`'s: the table contains a collection of foreign keys, each of the foreign keys points back to its table. When the local variables referencing pointing to these objects get unset, there are still references from the FKs to the table, so from the garbage collector's standpoint, the table refcount is greater than zero, so the graph doesn't need to be destructed.

I didn't dig into this specific issue deeply but in https://github.com/doctrine/dbal/issues/4515, the same relationship between Connection and SchemaManager/ExpressionBuilder does lead to connection leaks.

**Comment by morozov** [comment]
Closed via #4742, #4743.
